fix history_d path for all possible cases

GuillaumeBrochu opened the pull request:
To continue the discussion here : https://github.com/geneweb/geneweb/issues/485

I know the code is not optimal, but what could I do to improve it?

fablhx commented:
Trigger AppVeyor CI

fablhx commented:
try with this code:

```ocaml
value history_d conf =
  let path =
    match p_getenv conf.base_env "history_path" with
    [ Some path when path <> "" -> path
    | None -> "" ]
  in
  let bname =
    if Filename.check_suffix conf.bname ".gwb" then conf.bname
    else conf.bname ^ ".gwb"
  in
  List.fold_left
    Filename.concat path [Util.base_path [] bname; "history_d"]
;
```

GuillaumeBrochu commented:
I think this path that you proposed yesterday doesn't make sense (`path` and `Util.base_path` are mutually exclusive ...):
`List.fold_left Filename.concat path [Util.base_path [] bname; "history_d"]`

However, I have used a part of your suggestion to better manage the possible cases for `history_path`and I have updated the commit.

<s> As the variable `history_path` can be set differently for each base using the `.gwf` file, I have also updated the `a.gwf `file to avoid confusion. And in the case that a database administrator would like to propose a common default `history_path` to all database users through a default `.gwf` file, I added the base name as a sub directory to `history_path` to avoid mixing the histories of two different bases:
`Some path when path <> "" -> Filename.concat path bname`</s>

Compiled and tested on Linux. I got the expected result in all cases.

What do you think?

fablhx commented:
Actually I don't understand what is wrong at the moment. Also I don't understand your comment about the administrator

> As the variable history_path can be set differently for each base using the .gwf file, I have also updated the a.gwffile to avoid confusion

It can *only* be set per database since it's a database configuration.

GuillaumeBrochu commented:
Ah.. sorry. I detailled it here (https://github.com/geneweb/geneweb/issues/485#issue-191896980), but I forgot to repeat this important detail again.

The current code in the master branch works well EXCEPT if these three conditions are met:
(1) if gwd is launched with the -bd argument AND 
(2) if this -bd argument is <s>an absolute path (or contains "..")</s> different from `.` AND
(3) if history_path parameter is set in the .gwf file

This looks like a VERY particular case, but it's actually my usual test configuration with my test scripts :-) 

In that particular case, the current code in the master branch concatenate two absolute paths, `path`  and `Util.base_path []` (that's why I said that they were mutually exclusive...):
https://github.com/geneweb/geneweb/blob/ca236c2b7631a32af4bbb1c6a7123c4e017f1f71/src/history_diff.ml#L41-L42

This being said, this made me realized that my fix is still not good, so I'll prepare a new one.

GuillaumeBrochu committed:
fix history_d path for ALL possible cases

GuillaumeBrochu commented:
This new commit fixes the problem when using the -bd argument and gives the same result as previously for the other cases.

And forget my previous comment about the .gwf file and administrator.

GuillaumeBrochu commented:
This one is ready to be reviewed, again. I think this last iteration should be the final one.

GuillaumeBrochu commented:
This one is ready to be reviewed, again. I think this last iteration should be the final one.

fablhx commented:
I will look carefully at this PR, but only after this work (https://github.com/geneweb/geneweb/tree/base-architecture-rebase) is done.
In order not to leave you without any explanation, gw base directory is sometimes (too many time) complicated, allowing several (too much) configuration possible. This simplification will allow one configuration and one only.

GuillaumeBrochu commented:
OK. I will close it for now, since the solution will probably be different after your change, and I will leave a note into #485 (that is about the same problem).
